### PR TITLE
Isolate landed travel intent handling

### PR DESCRIPTION
**Pull request by ahmdkaml**
This is mainly for review/feedback on the direction rather than a complete implementation.

I traced the map travel flow from the client: the map sends a TravelIntent, which reaches Dispatch and then HandleTravel.

I intercepted TravelIntent at the dispatch boundary and wrapped the existing handling so the new case can be isolated without changing the current HandleTravel behavior for its other callers.

The idea is to handle the landed-on-planet → map travel case separately, since that case needs to become a space transit rather than immediately resetting/loading the destination world.

Next, I can use EnterSpace for the launch and connect it to the existing travel/landing flow so the client gets the proper take-off → space transit → landing sequence.

I held off on implementing the actual server-side transit sequence because I wanted to confirm that this direction matches the intended design before adding more code.

Sorry for the slow response — I had something holding me up for a bit.

If this direction looks right, I can finish the server-side portion in a follow-up commit and then open a separate PR for the Unity client changes.

### CODE REVIEW

**Comment by marceld23** [comment]
Thanks @ahmdkaml  — the direction is right and matches the sketch in #1614: intercept the landed-ship case, `EnterSpace` with a pending destination, then `LandOnBody`. A few things before you build the transit on it:

1. **Gate on "not in space", not on the body kind.** `session.CurrentLocationId` keeps the body id while the ship is flying (`EnterSpace` leaves it as the flight anchor), so `IsOnLandableBody` is also true for a ship in flight. Use `!InSpace(session.State.PlayerId)` — that is the actual "parked on a surface" signal.
2. **Keep the instant path** for `session.Spectating` and the admin `/tp` / `/goto` callers (`adminBypass`), as listed in the issue.
3. **Run the existing gates before take-off:** destination exists, `FreeSpaceFlight`, the Instant Travel rule, jump generator / jump lane, pad claim, plus the new cross-system "already landed there" check. A reject must happen while the player is still on the ground. Easiest is to split `HandleTravel` into a validate step and a landing step and let the transit share the first one instead of duplicating it. Note `HandleTravel` calls `Serve(session)` first — the wrapper needs that too before touching `_ship`.
4. **Right now the new branch swallows the intent silently**, so please keep the PR as a Draft until the transit lands; merged as-is every map travel from a surface would do nothing.

Server half first with tests for the three cases and the two gates is fine, client PR afterwards. No rush.

**Comment by ahmdkaml** [comment]
I think I have the case 1 flow narrowed down now. My current implementation is:

- Keep `TravelIntent` as the initial request.
- For a normal landed-ship transit, validate everything server-side while the player is still on the ground.
- Store the validated destination in `session.PendingTransitBodyId`.
- Send `SpaceState` with `AutomaticTransit = true` so the client knows this is an automatic transit and should play the launch animation.
- When the launch animation finishes, the client sends the existing `LeaveSpaceIntent` rather than introducing another protocol message.
- `HandleLeaveSpace` consumes and clears `PendingTransitBodyId`, then calls the existing `HandleTravel(...)` directly to complete the landing.
- Leave the normal `LeaveSpace` behavior untouched for ordinary space flight.

This gives case 1 a simple flow:

`TravelIntent → SpaceState(AutomaticTransit) → launch animation → LeaveSpaceIntent → HandleTravel → landing`

Before I proceed further, does this match the intended architecture for #1614? In particular, is using the existing `LeaveSpaceIntent` as the client’s “launch animation finished / ready to continue” signal the approach you want?

**Comment by ahmdkaml** [comment]
I can see the issues in the test but it's my first time doing a change that touches so many files. so I a little stressed. for the things u mentioned I covered some of them not all I will go back to them. but I need to know if the current direction is ok.
I can trigger a leavespace msg instead of a new msg entirely there is a lot of decisions I can't make

**Comment by ahmdkaml** [comment]
The current implementation covers case 1. For the remaining cases, my plan is:

- **Case 2:** send a signal from the server indicating that the client should play the space-flight animation. I think this can reuse `AutomaticTransit` rather than introducing a second state/message.
- The client plays the animation and then sends the existing `LeaveSpaceIntent`, which lets the server continue through the existing landing flow.
- **Case 3:** the same animation signal can be used, but after the animation the player remains in space/flight instead of sending `LeaveSpaceIntent`. This should be straightforward to distinguish based on the server-provided state.

So the idea is to keep the protocol small and reuse `AutomaticTransit` as the indication that an automatic travel animation needs to be played, while the subsequent client action determines whether we land or remain in flight.

Does this approach match the intended design?

Also, if this is the right direction, how would you like the tests structured for cases 1–3?

**Comment by marceld23** [comment]
Good progress @ahmdkaml — the gate, the validate/land split and the pending destination are exactly what I had in mind. Three things need changing before this can go further, and one answer to your question:

**Your question: no, please don't reuse `LeaveSpaceIntent`.** While a transit is pending, every `LeaveSpaceIntent` gets hijacked — the pad picker (`SendLeaveSpace(body, padIndex)` in `SpaceView`), the "Land" button in the ship menu, EVA landings, the `ScreenshotDirector`. Add a tiny dedicated intent (e.g. `TransitLaunchDoneIntent`, empty body, registered in `NetCodec` — your #979 guard will remind you) and handle it next to `HandleLeaveSpace`. Also add the server-side timeout from the issue: if the signal never arrives (disconnect mid-launch, hidden WebGL tab), the server lands the player itself after the launch length.

1. **CI is red: 9 tests fail** (`LanPlaytestRegressionTests` ×7, `BaseLifeTests` ×2), and `QuickTravel_IsGatedByInstantTravel_AndLandedHistory` is commented out. Root cause is the same: `QuickTravelForTest` now goes through the transit, so a landed test player ends up in space instead of on the target. Keep `QuickTravelForTest` on the instant `HandleTravel` (it is the "instant" helper) and add a second test entry for the transit — start it, assert the player is in space with the pending body set, send the done-signal, assert the landing. Please never comment a test out; adapt it.
2. **Client loop:** the new block in `UpdateSequence(rising)` fires `SendLeaveSpace()` every frame once `_seq >= SeqDuration` because the phase never changes. Latch it (send once, then hold). And the landing half is still a hard cut: after the server lands you, the client should run `Phase.Landing` toward the destination before the `WorldReset` takes over — that is the visible half of #1614. I'll do a local Unity build before merging anyway (CI does not compile `client/Assets`).
3. **Pad claimed twice:** `AllowNormalTravel` claims the chosen pad on the ground, then `HandleTravel` on arrival claims again with `PadIndex = -1`, so the player's choice is lost. Either skip the second claim when the transit already holds one, or carry the pad index in the pending state.

Still open from the issue: the cross-system "already landed there" gate must hold even when the Instant Travel rule is on (today it only bites when the rule is off), and case 3 (`HyperjumpToSystem` from a surface with `skipLaunch: false`). Both fine as follow-up commits. Small nits: `_galaxy.FindBody(...)!` next to `_galaxy?.` in the same block, and a stray blank line at the top of `HandleTravel`. Please mark the PR as Draft until CI is green.

**Comment by ahmdkaml** [comment]
only the tests remain for the server side now right?

**Comment by marceld23** [comment]
@ahmdkaml Almost — the tests are the last big item, but five fixes first (all small):

1. **`AutomaticTransit` gets stuck.** The timeout path in `TickEnvironment` clears `PendingTransitBodyId` but not `session.AutomaticTransit`. The next ordinary launch then sends `SpaceState.AutomaticTransit = true`, and the client never reaches `Phase.Cruise` (your new block returns before it) — the player hangs in the launch sequence. Clear the flag wherever the pending transit ends (launch-done, timeout), and defensively in `EnterSpace` when no transit is pending.
2. **Case 3 has the same bug, permanently.** `HyperjumpToSystem` from a surface sets `AutomaticTransit = true` but still calls `EnterSpace(skipLaunch: true, …)`, so no launch plays, no launch-done ever arrives, and the flag stays set forever. It should be `skipLaunch: !wasLanded` (the issue: take-off first, then the warp).
3. **Pad leak on reject.** The new cross-system "already landed there" check runs *after* `AllowNormalTravel`, which has already claimed a pad on the destination — a rejected jump leaves a pad occupied on a world the player never reaches. Move that check before the claim (look up origin/destination first, then call `AllowNormalTravel`).
4. **Pad still claimed twice.** The landing goes through `LandOnBody` → `HandleLeaveSpace` → `HandleTravel` with `PadIndex = -1`, so the pad the player picked on the ground is lost and a second one is claimed. Store the pad index with the pending transit (e.g. `PendingTransitPadIndex`) and pass it into the landing.
5. **Timeout: 2.0 s is tight.** The client launch is 1.6 s (`SeqDuration`) plus network, so the fallback often wins the race against the real signal. Harmless in effect, but a fallback should sit well behind the normal path — 6 s or so.

Housekeeping: `dotnet format` is red (trailing whitespace at GameServer.cs 950 / 3403–3404, NetCodecTests.cs 288 / 290, NetworkClient.cs 557 — run `dotnet format` before pushing), please restore the three-line comment you removed above the Instant Travel gate, put the blank lines back where they were (`HandleTravel` top, after the jump-generator check, in `HandleLeaveSpace`), and give `TransitLaunchDoneIntent` a proper doc sentence. The `Register(236, …)` line belongs at the end of the list after 227, not in the 105–107 block.

Then the tests from the issue: same-system transit (in space + pending body → launch-done → landed), cross-system transit, cross-system to a never-landed body rejected with no pad left claimed, flying-ship jump unchanged, and the timeout landing. Friday still looks fine.

**Comment by ahmdkaml** [comment]
this address most of the required change no review is needed u can wait until the tests to review them all at once if there is any requirement about the test or if these aren't addressing the issue I am writing the test and will respond immediately to any comment.
this wasn't clear as well "and defensively in EnterSpace when no transit is pending."

**Comment by marceld23** [comment]
@ahmdkaml Agreed — I'll review everything in one pass once the tests are in. I checked the clean-ups commit anyway: all five points are addressed. Only housekeeping left: `dotnet format` is still red on two files (NetworkClient.cs 557, NetCodecTests.cs 288/290), the three-line comment above the Instant Travel gate, and the `Register(236)` placement — fold those into the tests commit.

"Defensively in EnterSpace" was worded badly, sorry. What I meant: a *manual* launch from the helm is never a transit, so `HandleEnterSpace` (the `EnterSpaceIntent` handler, not `EnterSpace` itself — case 3 sets the flag right before calling that) should clear `session.AutomaticTransit` and `PendingTransitBodyId` before it launches. Two lines. Now that both end paths clear the flag it's only a belt-and-braces guard against a stale flag leaking into a `SpaceState`, but cheap enough to keep.

Plan from here: once the tests and `dotnet format` are green I'll do the local Unity build (CI doesn't compile `client/Assets`) and merge this PR as the server half. That way your client PR starts from a fresh main instead of chasing it — main has moved a fair bit since your last merge.

One question so I can set a realistic date for the client half: roughly how many hours can you put in over the next days? No wrong answer, I just want the date to match your calendar rather than my guess.

**Comment by marceld23** [comment]
@ahmdkaml I did the local Unity build of your branch tonight — the client half doesn't compile (9 errors, all in your `UpdateSequence` block):

- `Game.SpaceState` doesn't exist. `GameBootstrap` never keeps the `SpaceState` message; it latches single values on entry (see `SpaceSkipLaunch` at line ~620 and the `SpaceStateReceived` handler at ~2187).
- `_transitLaunchDone` is never declared.

This 5-line fix builds clean here (0 `error CS`):

```diff
--- a/client/Assets/BlocksBeyondTheStars/Scripts/GameBootstrap.cs
+++ b/client/Assets/BlocksBeyondTheStars/Scripts/GameBootstrap.cs
@@ -620 +620,2 @@
         public bool SpaceSkipLaunch { get; private set; }    // entered space already airborne (helm) → no take-off anim
+        public bool SpaceAutomaticTransit { get; private set; } // #1614: this flight is an automatic landed-ship transit (launch → signal → landing)
@@ -2189 +2190,2 @@
                     SpaceSkipLaunch = m.SkipLaunch; // latched on entry only (later updates don't re-trigger Enter)
+                    SpaceAutomaticTransit = m.AutomaticTransit; // #1614: latched with it — the launch sequence reads it once it ends
--- a/client/Assets/BlocksBeyondTheStars/Scripts/SpaceView.cs
+++ b/client/Assets/BlocksBeyondTheStars/Scripts/SpaceView.cs
@@ -333 +333,2 @@
         private bool _hyperjumping; // a hyperspace jump is tearing down the view (warp covers it, no landing)
+        private bool _transitLaunchDone; // #1614: the automatic transit's launch-done signal was sent (once per flight)
@@ -988 +989 @@
-                if (Game.SpaceState.AutomaticTransit)
+                if (Game.SpaceAutomaticTransit)
@@ -2538 +2539,2 @@
             _hyperjumping = false;
+            _transitLaunchDone = false;
```

Please apply it with the tests commit. And for the client PR later: open `client/` in Unity once before pushing — the console shows compile errors right away, and CI never compiles `client/Assets`, so the build here is the only check. If you'd rather I push this fix onto your branch, say so and I will.

**Comment by ahmdkaml** [comment]
Yes, please push the client fixes. I’m more focused on the tests right now, so don’t worry — for any future contribution, if one comes up, I’ll have Unity set up locally and compile/test the client before pushing. Since I also play the game, I’ll manually test things in-game before pushing as well, which should keep your overhead to a minimum.

I just didn’t want to start dealing with Unity right now because I’m focusing on the server side and getting this half finished properly.

I’ll handle the remaining server fixes and run the formatter, then push. That way we can get this server half clear and ready for you to merge.

**Comment by marceld23** [comment]
Pushed the compile fix to your branch as dcc8006e (client only, two files). Please `git pull` before you commit the tests so your push fast-forwards. Thanks for the note on future client work — that plan sounds right.

**Comment by ahmdkaml** [comment]
oh I forgot the safegaurd no problem I will add it after this pass u can check the tests in the meanwhile.
I think I can make it before the deadline what do u think?

**Comment by ahmdkaml** [comment]
I will add these two lines and go to the convenience store if everything go red u can merge the previous commit or just delete them
session.AutomaticTransit = false;
        session.PendingTransitBodyId = null;

**Comment by marceld23** [comment]
@ahmdkaml You've made it — the server half is done. The five tests are exactly what I asked for and they go through the real loopback transport, which is the right level. The safeguard commit is in; as soon as CI is green on it I'll merge this PR.

Two small things you can pick up in the client PR later, no need to touch this one: a test for case 3 (`HyperjumpToSystem` from a surface should set `AutomaticTransit` and launch with `skipLaunch: false`, and a launch-done should clear it), and `Register(236)` still sits in the 105–107 block instead of after 227.

Enjoy the convenience store. When you're back: how many hours can you spare over the next two weeks? Then I'll put a date on the client half.

**Comment by ahmdkaml** [comment]
I can't really give an exact number since college starts in 8 days, so I expect my schedule to shift around a bit. For now, I'd estimate around 2–3 hours a day, roughly 4 days a week — so around 8–12 hours a week, probably more rather than less. Once college starts, I'll have a better idea of what my schedule actually looks like.
@marceld23
